---
title: "Visualizing ICA results with R package icreport"
author: "Jin-Hyun Ju"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

The `icreport` package provides functions that perform independent component analysis on a given data set and can subsequently generate a compact report. 

## Quick Start

### 1.Loading and Preparing an Example Dataset

If you already know what you want to use the package for, follow this simple example to get started!

1) The package is not yet published on bioconductor or CRAN, so the best way to install the package is directly from github. Installing the package through the function ```install_github``` from the package ```devtools```.

```{r Install icreport, eval = FALSE}
install.package("devtools") # in case you don't have devtools already

devtools::install_github("jinhyunju/icreport") #installing icreport
```

2) Installing and loading an example dataset that is available through bioconductor.

- Here we are going to use the bladderbatch dataset that was used in the tutorial for surrogate variable analysis. (link to sva tutorial)

- You can use your own dataset, it just needs to be a matrix which has the dimension of (gene x samples). A dataframe with covariate information is optional with dimensions (samples x covariates). 

```{r Install datasets, eval = FALSE}
#source("http://bioconductor.org/biocLite.R")
#biocLite("golubEsets")

source("http://bioconductor.org/biocLite.R")
biocLite("bladderbatch")
```

- Onething that you have to watch out for is that the rownames of ```info.df``` have to match the column names of the ```phenotype.mx```. They don't necessarily have to be in the same order but they should have names that overlap. 

```{r Load bladder batch, eval = FALSE, echo= TRUE }
library(bladderbatch)
data(bladderdata)

info.df <- pData(bladderEset)    # dimensions 57 x 4 (57 samples, 4 features) is the info.df 
expr <- exprs(bladderEset)     # dimensions 22283 x 57 (22283 probes, 57 samples)


info.df$batch <- as.factor(info.df$batch) #changing the numerical column into factors
check.covars <- c("batch","outcome")

```

### 2.Running ICA on the expression dataset. 

2-1) Things that you need:

- phenotype.mx is the expression matrix with genes in rows and samples in columns (dimension = genes x samples)

- info.df is the sample information table with the covariates in columns (dimension = samples x covariates)

- check.covars is a vector that contains the names of the covariates (column names of info.df) that should be tested for correlation. 

- For information regarding other advanced options please use ```?gene_expr_ica```.

2-2) In case you have covariates:

```{r ICA w covars, eval = FALSE}
library(icreport)
ica_result <- gene_expr_ica(phenotype.mx = expr, info.df = info.df, 
                            check.covars = check.covars)

```

2-3) In case you don't have any covariates:

```{r ICA wo covars, eval = FALSE}
library(icreport)
ica_result <- gene_expr_ica(phenotype.mx = expr)

```

2-4) Creating a report file in the current working directory

- ica.result is the result generated by the ```gene_expr_ica()``` function.

- prefix specifies the name of the html report file. 

```{r ICA report, eval = FALSE}
report2me(input = ica_result, mode = "ica", prefix = "Test_ICA_Report")
```

Troubleshooting if report2me() fails

- report2me() generates an html report using ```knitr``` which depends on ```pandoc```. If you are running this in Rstudio it will most likely work since Rstudio comes with pandoc, but if you are running this on plain R you might have to install pandoc seperately. Here is a link which shows the details for platform specific pandoc installations. https://github.com/rstudio/rmarkdown/blob/master/PANDOC.md

### 3.Running PCA on the expression dataset. 

3-1) In case you have covariates:

```{r PCA w covars, eval = FALSE}

pca_result <- gene_expr_pca(phenotype.mx = expr, info.df = info.df, 
                            check.covars = check.covars)

```

3-2) In case you don't have any covariates:

```{r PCA wo covars, eval = FALSE}

pca_result <- gene_expr_pca(phenotype.mx = expr)

```

3-3) Generating a report:

- pca.result is the result generated by the ```gene_expr_pca()``` function.

- prefix specifies the name of the html report file. 

```{r PCA report, eval = FALSE}
report2me(input = pca_result, mode = "pca", prefix = "Test_PCA_Report")
```


## Introduction

Gene expression analyses with RNA sequencing or microarray produce extremely high dimensional datasets containing approximately 20,000 or more features per sample. Inspecting such large datasets is no easy task. Heatmaps or Principal Component Analysis (PCA) are commonly used for visual representation of the data to reveal any hidden structures. Independent component analysis (ICA) is a method that can give us valuable insight about the hidden structure of the data. 

